### PR TITLE
RabbitMq configuration

### PR DESCRIPTION
**Pull request by SergioSilvaLeon**
Configured RabbitMq to communicate with the service via a simple string for the moment.

### CODE REVIEW

File: event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.springframework.amqp.rabbit.annotation.RabbitHandler;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+
+@RabbitListener(queues = "productQueue")
+public class ProductService {
+
+    @RabbitHandler
+    public void receive(String in) {
+        System.out.println("[x] Received '" + in + "'");
```
Comment (wombstrat) [inline event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java]: Use Loggers, instead of standard output

File: pom.xml [hunk 2]
```diff
@@ -13,4 +13,14 @@
         <module>event-subscription-product</module>
     </modules>
 
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-starter-amqp</artifactId>
+                <version>2.0.4.RELEASE</version>
```
Comment (wombstrat) [inline pom.xml]: Try to put this on a <property/>

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import com.nearsoft.app.rabbitMq.Producer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ProductController {
+
+    @Autowired
+    Producer producer;
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: Use constructor based dependency injection.

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.nearsoft.app.controller;
+
+import com.nearsoft.app.rabbitMq.Producer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ProductController {
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: Is needed for this Controller to be public

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.amqp.core.Queue;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class RabbitConfig {
+
+    @Bean
+    public Queue queue() {
+        return new Queue("productQueue");
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: Put _productQueue_ either in a properties file or define it as a constant

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    }
+
+    @Bean
+    public Producer sender() {
+        return new Producer();
+    }
+
+    @Bean
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: Isn't already ProductService as a spring bean?

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+@Configuration
+public class RabbitConfig {
+
+    @Bean
+    public Queue queue() {
+        return new Queue("productQueue");
+    }
+
+    @Bean
+    public Producer sender() {
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: Same, isn't already Producer class defined as a Spring Bean

File: event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.nearsoft.service;
+
+import org.springframework.amqp.rabbit.annotation.RabbitHandler;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+
+@RabbitListener(queues = "productQueue")
+public class ProductService {
+
+    @RabbitHandler
+    public void receive(String in) {
```
Comment (wombstrat) [inline event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java]: Use more descriptive names for arguments.

File: event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.nearsoft.service;
+
+import org.springframework.amqp.rabbit.annotation.RabbitHandler;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+
+@RabbitListener(queues = "productQueue")
```
Comment (wombstrat) [inline event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java]: Use a constant here.

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+public class Producer {
+
+    @Autowired
+    private RabbitTemplate rabbitTemplate;
+
+    @Autowired
+    private Queue queue;
+
+    public void produceMessage() {
+        // Test sender works
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: Avoid this kind of comments, if something is needed to be documented use javadoc standards.

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        String message = "RabbitMq Message";
+        this.rabbitTemplate.convertAndSend(queue.getName(), message);
+        System.out.println("[x] Sent '" + message + "'");
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: Logger again. Check also String#format

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    @Autowired
+    private RabbitTemplate rabbitTemplate;
+
+    @Autowired
+    private Queue queue;
+
+    public void produceMessage() {
+        // Test sender works
+        String message = "RabbitMq Message";
+        this.rabbitTemplate.convertAndSend(queue.getName(), message);
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: if there's no local variable also named `rabbitTemplate`, why is `this` needed?

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+    @Autowired
+    private RabbitTemplate rabbitTemplate;
+
+    @Autowired
+    private Queue queue;
+
+    public void produceMessage() {
+        // Test sender works
+        String message = "RabbitMq Message";
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: Will all messages be hardcoded?

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class Producer {
+
+    @Autowired
```
Comment (wombstrat) [inline web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: Again, use constructor based injection.

File: pom.xml (hunk 2 above)
Comment (SergioSilvaLeon) [reply pom.xml]: @wombstrat What do you exactly mean?

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java (hunk 4 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: @wombstrat I'm sorry may you please elaborate more?

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java (hunk 6 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: When I remove the @bean with productService the application seems to stop consuming.

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java (hunk 4 above)
Comment (wombstrat) [reply web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: why we should the `public` access modifier for this class?

File: pom.xml (hunk 2 above)
Comment (wombstrat) [reply pom.xml]: Sorry I meant: "Try to put this on a `</property>` tag

File: event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java (hunk 8 above)
Comment (SergioSilvaLeon) [reply event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java]: done

File: event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java (hunk 1 above)
Comment (SergioSilvaLeon) [reply event-subscription-product/src/main/java/com/nearsoft/service/ProductService.java]: done

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java (hunk 4 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: done

File: web-app/src/main/java/com/nearsoft/app/controller/ProductController.java (hunk 3 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/controller/ProductController.java]: donde

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java (hunk 14 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java (hunk 10 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java (hunk 13 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java (hunk 12 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java (hunk 11 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/Producer.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java (hunk 5 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: done

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java (hunk 7 above)
Comment (SergioSilvaLeon) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: removed @Bean and Producer method

**Comment by SergioSilvaLeon** [comment]
Hi, @wombstrat I updated the changes you've request. I hope it all seems in order

File: web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java (hunk 6 above)
Comment (wombstrat) [reply web-app/src/main/java/com/nearsoft/app/rabbitMq/RabbitConfig.java]: If I'm not seeing and outdated file, the ProductService class is annotated with `@Component`, which defines it as a Bean for Spring, is the @ComponentScan finding it? (ProductService)

**Pull request merged**